File: app/services/finance_calculations_service.rb
class FinanceCalculationsService
  def call(income = 0, salary = 0)
    income_tax = calculate_tax(income)
    national_insurance = calculate_national_insurance(salary)
    student_loan = calculate_student_loan_repayment(income)

    disposable_income = income - (income_tax + national_insurance + student_loan)

    return {
      income_tax: income_tax,
      national_insurance: national_insurance,
      student_loan: student_loan,
      disposable_income: disposable_income
    }
  end

  def calculate_tax(income)
    return 0 if income.nil?
    # Define tax brackets and rates
    personal_allowance = 12570
    basic_rate_limit = 50270
    higher_rate_limit = 125140
    tax_basic = 0.20
    tax_higher = 0.40
    tax_super = 0.45
    # Calculate tax based on income
    if income <= personal_allowance
      tax = 0
    elsif income <= basic_rate_limit
      tax = (income - personal_allowance) * tax_basic
    elsif income <= higher_rate_limit
      tax = (basic_rate_limit - personal_allowance) * tax_basic + (income - basic_rate_limit) * tax_higher
    else
      tax = (basic_rate_limit - personal_allowance) * tax_basic + (higher_rate_limit - basic_rate_limit) * tax_higher + (income - higher_rate_limit) * tax_super
    end
    tax
  end

  def calculate_national_insurance(salary)
    return 0 if salary.nil?

    # Annual thresholds for national insurance contributions
    annual_lower_threshold = 12570  # Example: £12,570
    annual_upper_threshold = 50270  # Example: £50,270
    annual_additional_threshold = 96700  # Example: £96,700

    # Calculate NI based on annual salary
    if salary <= annual_lower_threshold
      ni = 0
    elsif salary <= annual_upper_threshold
      ni = (salary - annual_lower_threshold) * 0.12  # 12% rate between lower and upper threshold
    elsif salary > annual_upper_threshold
      ni = (annual_upper_threshold - annual_lower_threshold) * 0.12 + (salary - annual_upper_threshold) * 0.02  # 2% on everything above upper threshold
    end

    ni
  end

  def calculate_student_loan_repayment(income)
    return 0 if income.nil?
    # Define tax brackets and rates
    payment_threshold = 27295
    tax_basic = 0.09
    # Calculate tax based on income
    if income <= payment_threshold
      studentloan = 0
    else
      studentloan = (income - payment_threshold) * tax_basic
    end
    studentloan
  end
end

File: app/javascript/finance_calculations_service.js
export class FinanceCalculationsService {
  call(income = 0, salary = 0) {
      const incomeTax = this.calculateTax(income);
      const nationalInsurance = this.calculateNationalInsurance(income);
      const studentLoan = this.calculateStudentLoanRepayment(income);

      const disposableIncome = income - (incomeTax + nationalInsurance + studentLoan);

      return {
          incomeTax: incomeTax,
          nationalInsurance: nationalInsurance,
          studentLoan: studentLoan,
          disposableIncome: disposableIncome
      };
  }

  calculateTax(income) {
      if (income === null) return 0;

      const personalAllowance = 12570;
      const basicRateLimit = 50270;
      const higherRateLimit = 125140;
      const taxBasic = 0.20;
      const taxHigher = 0.40;
      const taxSuper = 0.45;

      let tax = 0;
      if (income <= personalAllowance) {
          tax = 0;
      } else if (income <= basicRateLimit) {
          tax = (income - personalAllowance) * taxBasic;
      } else if (income <= higherRateLimit) {
          tax = (basicRateLimit - personalAllowance) * taxBasic + (income - basicRateLimit) * taxHigher;
      } else {
          tax = (basicRateLimit - personalAllowance) * taxBasic + (higherRateLimit - basicRateLimit) * taxHigher + (income - higherRateLimit) * taxSuper;
      }

      return tax;
  }

  calculateNationalInsurance(salary) {
    if (salary === null) return 0;

    // Annual thresholds for national insurance contributions
    const annualLowerThreshold = 12570;  // Adjusted to match annual figures, e.g., £12,570
    const annualUpperThreshold = 50270;  // Adjusted to match annual figures, e.g., £50,270
    const annualAdditionalThreshold = 9670;  // Adjusted to match annual figures, e.g., £96,700

    let ni = 0;
    if (salary <= annualLowerThreshold) {
        ni = 0;
    } else if (salary <= annualUpperThreshold) {
        ni = (salary - annualLowerThreshold) * 0.12;  // 12% between lower and upper threshold
    } else if (salary <= annualAdditionalThreshold) {
        ni = (annualUpperThreshold - annualLowerThreshold) * 0.12 + (salary - annualUpperThreshold) * 0.02;  // 2% on everything above the upper threshold
    } else {
        // If salary exceeds additional threshold
        ni = (annualUpperThreshold - annualLowerThreshold) * 0.12 + (annualAdditionalThreshold - annualUpperThreshold) * 0.02;
    }

    return ni;
  }

  calculateStudentLoanRepayment(income) {
      if (income === null) return 0;

      const paymentThreshold = 27295;
      const taxBasic = 0.09;

      let studentLoan = 0;
      if (income > paymentThreshold) {
          studentLoan = (income - paymentThreshold) * taxBasic;
      }

      return studentLoan;
  }
}
